src/Microsoft.Health.Fhir.Shared.Core/Features/Search/Parameters/SearchParameterValidator.cs

```diff
@@ -0,0 +1,120 @@
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using EnsureThat;
+using FluentValidation.Results;
+using Hl7.Fhir.Model;
+using Microsoft.Health.Extensions.DependencyInjection;
+using Microsoft.Health.Fhir.Core;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Features.Definition;
+using Microsoft.Health.Fhir.Core.Features.Operations;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Features.Security;
+using Microsoft.Health.Fhir.Core.Features.Security.Authorization;
+using Microsoft.Health.Fhir.Core.Features.Validation;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Shared.Core.Features.Search.Parameters
+{
+    public class SearchParameterValidator : ISearchParameterValidator
+    {
+        private readonly Func<IScoped<IFhirOperationDataStore>> _fhirOperationDataStoreFactory;
+        private readonly IFhirAuthorizationService _authorizationService;
+        private readonly ISearchParameterDefinitionManager _searchParameterDefinitionManager;
+
+        private const string HttpPostName = "POST";
+        private const string HttpPutName = "PUT";
+        private const string HttpDeleteName = "DELETE";
+
+        public SearchParameterValidator(
+            Func<IScoped<IFhirOperationDataStore>> fhirOperationDataStoreFactory,
+            IFhirAuthorizationService authorizationService,
+            ISearchParameterDefinitionManager searchParameterDefinitionManager)
+        {
+            EnsureArg.IsNotNull(fhirOperationDataStoreFactory, nameof(fhirOperationDataStoreFactory));
+            EnsureArg.IsNotNull(authorizationService, nameof(authorizationService));
+            EnsureArg.IsNotNull(searchParameterDefinitionManager, nameof(searchParameterDefinitionManager));
+
+            _fhirOperationDataStoreFactory = fhirOperationDataStoreFactory;
+            _authorizationService = authorizationService;
+            _searchParameterDefinitionManager = searchParameterDefinitionManager;
+        }
+
+        public async Task ValidateSearchParamterInput(SearchParameter searchParam, string method, CancellationToken cancellationToken)
+        {
+            if (await _authorizationService.CheckAccess(DataActions.Reindex) != DataActions.Reindex)
+            {
+                throw new UnauthorizedFhirActionException();
+            }
+
+            // check if reindex job is running
+            using (IScoped<IFhirOperationDataStore> fhirOperationDataStore = _fhirOperationDataStoreFactory())
+            {
+                if (await fhirOperationDataStore.Value.CheckActiveReindexJobsAsync(cancellationToken))
+                {
+                    throw new JobConflictException(Resources.ChangesToSearchParametersNotAllowedWhileReindexing);
```

Comment:
Can we pass address or id of reindex job to user so user can terminate or monitor that job?

Reply:
Updated.